src/renderers/shared/fiber/ReactFiberClassComponent.js

```diff
@@ -73,6 +74,28 @@ module.exports = function(scheduleUpdate : (fiber: Fiber, priorityLevel : Priori
     },
   };
 
+  function checkShouldComponentUpdate(workInProgress, oldProps, newProps, newState) {
+    if (oldProps === null ||
+        (workInProgress.updateQueue && workInProgress.updateQueue.isForced)) {
+      return true;
+    }
+
+    const instance = workInProgress.stateNode;
+    if (typeof instance.shouldComponentUpdate === 'function') {
+      return instance.shouldComponentUpdate(newProps, newState);
+    }
+
+    const type = workInProgress.type;
+    if (type.prototype && type.prototype.isPureReactComponent) {
```

Comment:
I figured we start off with all of these being detected on demand and then we can create optimized paths for common patterns. E.g. for things that have shouldComponentUpdate, and things that are pure components and things that have no update life-cycles etc. by storing that in the tag like @acdlite said.